bring grids for (published) pages on par with workflows

martenson opened the pull request:
~~this is waiting on https://github.com/galaxyproject/galaxy/pull/16163, will stay draft until then~~

- unifies the component for published/personal pages grid
- reuses components from workflows grids
- converts some components to TS/compo, brings others closer to current standard
- brings new junit, API, and selenium test coverage


### personal grid
- before ![Galaxy](https://github.com/galaxyproject/galaxy/assets/1814954/11c37442-1c1f-498d-83ab-a1077cd5966a)
- after ![Galaxy___martenson](https://github.com/galaxyproject/galaxy/assets/1814954/6bc030e7-3485-4374-8508-8bfc92d95664)

### published grid
- before  ![Galaxy](https://github.com/galaxyproject/galaxy/assets/1814954/fbaa9ce7-a767-4b6c-8cf6-718d125bfc65)
- after ![Galaxy___martenson](https://github.com/galaxyproject/galaxy/assets/1814954/6a24e6cf-53e1-4906-8c37-12f658b9b1c8)

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

martenson committed:
add page dropdown for debug sharing

martenson committed:
change the test to not directly access comp function

martenson committed:
standardize indexfilter icon handling

martenson committed:
add new component for page index actions

martenson committed:
standardize icon handling in pagedropdown

martenson committed:
finish client changes to the page grid components

martenson committed:
add backend changed for pages grids

martenson committed:
juggle the enums in schema

see https://github.com/galaxyproject/galaxy/pull/15774 for context

update page =pdf export selenium test

update return type to reflect matches being returned

cleanup the test file

run make format

martenson committed:
adjust test_pages selenium test for new grid markup

jdavcs commented on lib/galaxy/managers/pages.py:
```diff
@@ -98,34 +121,93 @@ def __init__(self, app: MinimalManagerApp):
             query = query.filter(model.Page.user_id == payload.user_id)
 
-        total_matches = query.count()
+        if payload.search:
```
I know this method does only one thing, but it's still quite lengthy, and this new addition makes it even more so. Would you be opposed to breaking it up a bit for the sake of making it more readable (and testable down the road)? I think even factoring out the handling of `payload.search` would be an improvement.

martenson replied:
This one has a sister method in `managers/workflows.py` that has related logic and is really helpful with debugging. If I were to refactor it I'd include both and I am vary of increasing the scope of this PR. 

The whole method is also working on the same query object which in my opinion to some extent warranties its somewhat increased length.

jdavcs commented on lib/galaxy/webapps/galaxy/api/pages.py:
```diff
@@ -78,6 +80,19 @@
+    IndexQueryTag("title", "The pages's title."),
+    IndexQueryTag("slug", "The pages's slug.", "s"),
+    IndexQueryTag("tag", "The pages's tags.", "t"),
+    IndexQueryTag("user", "The pages's owner's username.", "u"),
```
`pages's` >> `page's`?

jdavcs commented on lib/galaxy/schema/schema.py:
```diff
@@ -1257,20 +1257,19 @@ class InvocationIndexQueryPayload(Model):
+    sort_desc: Optional[bool] = Field(default=False, title="Sort descending", description="Sort in descending order.")
+    search: Optional[str] = Field(default=None, title="Filter text", description="Freetext to search.")
+    limit: Optional[int] = Field(default=100, lt=1000, title="Limit", description="Maximum number of pages to return.")
+    offset: Optional[int] = Field(default=0, title="Offset", description="Number of pages to skip")
```
Nitpicking: drop period for consistent punctuation in description?

martenson committed:
correct grammar in desc

jdavcs commented on lib/galaxy_test/api/test_pages.py:
```diff
@@ -111,6 +112,10 @@ def test_index(self):
         create_response_json = self._create_valid_page_with_slug("indexpage")
         assert self._users_index_has_page_with_id(create_response_json)
 
+    def test_index_published_shared(self):
```
Could you rename this test to reflect that it's expected to produce a non-2xx response, and/or why that's expected? (like similar tests in the same module - `test_cannot_create...` or `test_400_on_invalid...`)

jdavcs reviewed (approved):
backend looks good!

martenson committed:
unify punctuation handling in pages api desc

martenson committed:
rename method to have a more proper name

itisAliRH commented on client/src/components/Page/PageDropdown.vue:
```diff
@@ -0,0 +1,85 @@
+
+function onDelete() {
+    const confirmationMessage = _l(`Are you sure you want to delete page`) + ` "${props.page.title}"?`;
+    if (window.confirm(confirmationMessage)) {
```
I think it would be better to utilize the modal instead of using the browser confirmation alert for deletion.

martenson replied:
@itisAliRH Do you mean a specific modal usage? Like our ui-modal.js component? I generally prefer the default for such simple use cases because it has well-defined and user-known display and shortcuts in all browsers.

itisAliRH replied:
@martenson I mean using bootstrap modal; e.g. [here](https://github.com/galaxyproject/galaxy/blob/dev/client/src/components/History/CurrentHistory/HistoryNavigation.vue#L161) it's a simple use case too but using a single, unified modal component is beneficial because it promotes consistency across the application.

martenson replied:
I like that, thanks for the suggestion!

itisAliRH commented on client/src/components/Page/PageIndexActions.vue:
```diff
@@ -0,0 +1,27 @@
+const props = defineProps<PageIndexActionsProps>();
+
+function create() {
+    window.location.assign(`${props.root}pages/create`);
```
Consider using composable `useRouter` instead to avoid a full app refresh.

itisAliRH commented on client/src/components/Page/PageList.vue:
```diff
@@ -0,0 +1,227 @@
+            v-if="rows >= perPage"
+            v-model="currentPage"
+            class="gx-pages-grid-pager"
+            v-bind="paginationAttrs"></b-pagination>
```
```suggestion
            v-bind="paginationAttrs" />
```

itisAliRH commented on client/src/components/Page/PageList.vue:
```diff
@@ -0,0 +1,227 @@
+                    :title="'Search more from ' + row.item.username"
+                    @click="appendTagFilter('user', row.item.username)"
+                    >{{ row.item.username }}</a
+                >
```
```suggestion
                    @click="appendTagFilter('user', row.item.username)">
                    {{ row.item.username }}
                </a>
```

itisAliRH reviewed (changes requested):
Overall, looks great to me👍
I've reviewed the front-end changes and left some comments.

martenson committed:
prettier than the prettier

martenson committed:
use router instead of window

martenson committed:
pretty

martenson committed:
wait for the render before setting form values

martenson committed:
refactor window.confirm to b-modal as suggested in a review

- drop the relevant jest tests since b-modal attaches outside of the wrapper and
is not findable to my knowledge
- could be likely replaced by a selenium test?

martenson committed:
re-add the jest unit tests with workadrounds for lazy load and portalling

this is what the actual component unit tests do
see https://github.com/bootstrap-vue/bootstrap-vue/blob/dev/src/components/modal/modal.spec.js#L233

martenson committed:
adjust test for new modal

martenson committed:
wait for render after page creation test method

martenson committed:
increase waiting time

martenson committed:
Merge branch 'dev' into more-pages

martenson committed:
adapt to new (conflicting?) navigation.yml

martenson committed:
Merge branch 'dev' into more-pages

 Conflicts:
	lib/galaxy_test/base/populators.py

martenson committed:
format

martenson committed:
attempt to fix selenium mby?

martenson committed:
attempt to please selenium diety

martenson committed:
revert pdf_export test changes, that broke things for no reason
add test_pages another attempt to please selenium deity with more waiting for no reason

davelopez commented on client/src/components/Page/PageIndexActions.vue:
```diff
@@ -0,0 +1,29 @@
+const props = defineProps<PageIndexActionsProps>();
+
+function create() {
+    router.push(`${props.root}pages/create`);
```
As I mentioned in https://github.com/galaxyproject/galaxy/pull/16209#issuecomment-1598327830 I think the issue is here:

```suggestion
    router.push("/pages/create");
```

The router should be smart enough to handle the prefixes so we don't have to use the `root` anymore. You can get rid of the props here too.

martenson committed:
remove custom prefix handling

thanks @davelopez
